catchup/service.go

```diff
@@ -866,38 +867,138 @@ func (s *Service) roundIsNotSupported(nextRound basics.Round) bool {
 	return true
 }
 
-func createPeerSelector(net network.GossipNode, cfg config.Local, pipelineFetch bool) *peerSelector {
-	var peerClasses []peerClass
+func createPeerSelector(net network.GossipNode, cfg config.Local, pipelineFetch bool) peerSelector {
+	var wrappedPeerSelectors []*wrappedPeerSelector
+	//TODO: Revisit this ordering
 	if pipelineFetch {
 		if cfg.NetAddress != "" && cfg.EnableGossipService { // Relay node
-			peerClasses = []peerClass{
-				{initialRank: peerRankInitialFirstPriority, peerClass: network.PeersConnectedOut},
-				{initialRank: peerRankInitialSecondPriority, peerClass: network.PeersPhonebookArchivalNodes},
-				{initialRank: peerRankInitialThirdPriority, peerClass: network.PeersPhonebookRelays},
-				{initialRank: peerRankInitialFourthPriority, peerClass: network.PeersConnectedIn},
+			wrappedPeerSelectors = []*wrappedPeerSelector{
+				{
+					peerClass: network.PeersConnectedOut,
+					peerSelector: makeRankPooledPeerSelector(net,
+						[]peerClass{{initialRank: peerRankInitialFirstPriority, peerClass: network.PeersConnectedOut}}),
+					priority:        peerRankInitialFirstPriority,
+					toleranceFactor: 3,
+					lastCheckedTime: time.Now(),
+				},
+				{
+					peerClass: network.PeersPhonebookRelays,
+					peerSelector: makeRankPooledPeerSelector(net,
+						[]peerClass{{initialRank: peerRankInitialFirstPriority, peerClass: network.PeersPhonebookRelays}}),
+					priority:        peerRankInitialSecondPriority,
+					toleranceFactor: 3,
+					lastCheckedTime: time.Now(),
+				},
+				{
+					peerClass: network.PeersPhonebookArchivalNodes,
+					peerSelector: makeRankPooledPeerSelector(net,
+						[]peerClass{{initialRank: peerRankInitialFirstPriority, peerClass: network.PeersPhonebookArchivalNodes}}),
+					priority:        peerRankInitialThirdPriority,
+					toleranceFactor: 10,
+					lastCheckedTime: time.Now(),
+				},
+				{
+					peerClass: network.PeersConnectedIn,
+					peerSelector: makeRankPooledPeerSelector(net,
+						[]peerClass{{initialRank: peerRankInitialFirstPriority, peerClass: network.PeersConnectedIn}}),
+					priority:        peerRankInitialFourthPriority,
+					toleranceFactor: 3,
+					lastCheckedTime: time.Now(),
+				},
 			}
 		} else {
-			peerClasses = []peerClass{
-				{initialRank: peerRankInitialFirstPriority, peerClass: network.PeersPhonebookArchivalNodes},
-				{initialRank: peerRankInitialSecondPriority, peerClass: network.PeersConnectedOut},
-				{initialRank: peerRankInitialThirdPriority, peerClass: network.PeersPhonebookRelays},
+			wrappedPeerSelectors = []*wrappedPeerSelector{
```

Comment:
For pipelinedFetch=true and relay=false, the priority order has changed.

Before:
1. `PeersPhonebookArchivalNodes`
2. `PeersConnectedOut`
3. `PeersPhonebookRelays`

After:
1. `PeersConnectedOut`
2. `PeersPhonebookRelays`
3. `PeersPhonebookArchivalNodes`

For pipelinedFetch=true and relay=true, there's also a swap between the priority ordering of `PeersPhonebookArchivalNodes` and `PeersPhonebookRelays`.

Is this intentional, and if so what's the rational here?

Reply:
I just talked this over with @algorandskiy; general thinking:

- If you are a relay, trying peers you are connected to first makes sense, followed by other relays, archival nodes
- If you are not a relay, the same order works
- I'm leaning towards dropping ConnectedIn to get rid of the relay cases (shrink this section from 4 to 2) - that's really the only diff between relays/non-relays and it's likely almost no benefit.

Will do a follow up commit here to address. Similarly for catchpoint catchup, doing relays then archival nodes.

Big question: is the hard-coded delay too long, too short, a good starting point, etc...